torch_ttnn/passes/lowering/to_tt_pass.py

```diff
@@ -555,6 +555,10 @@ def ReplaceMoreTtManually(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
                     permutation = [1, 0]
                     new_node = g.call_function(ttnn.permute, args=(args[0], permutation))
                     new_nodes.append(new_node)
+            elif node.target == torch.ops.aten.convolution.default:
+                if len(node.meta["val"].size()) == 4:
+                    new_node = g.call_function(ttnn.conv2d, args, node.kwargs)
```

Comment:
can't believe arguments match
https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_pybind.cpp#L34-L130

Reply:
I got this error.  Are we on the same page?  I'm trying remapping arguments in the following hours.
```
    def __call__(self, *function_args, **function_kwargs):
>       return self.function(*function_args, **function_kwargs)
E       TypeError: conv2d() takes 0 positional arguments but 9 were given

../tt-metal/ttnn/ttnn/decorators.py:327: TypeError
```

Aten is un(der)documented and hard to debug for everyone.  😅

Reply:
> I got this error.
Yes. Error indicates a mismatch.

> I'm trying remapping arguments in the following hours.
Got it 👍

Reply:
I already remapped the function arguments.  However, the compiler still reported incompatible function arguments.
```
E           TypeError: conv2d(): incompatible function arguments. The following argument types are supported:
E               1. (*, input_tensor: tt_lib.tensor.Tensor, weight_tensor: tt_lib.tensor.Tensor, device: tt_lib.device.Device, in_channels: int, out_channels: int, batch_size: int, input_height: int, input_width: int, kernel_size: List[int[2]], stride: List[int[2]], padding: List[int[2]], dilation: List[int[2]], groups: int, bias_tensor: Optional[tt_lib.tensor.Tensor] = None, conv_config: Optional[ttnn::operations::conv2d::Conv2dConfig] = None) -> Tuple[tt_lib.tensor.Tensor, int, int, tt_lib.tensor.Tensor, Optional[tt_lib.tensor.Tensor]]
E           
E           Invoked with: kwargs: input_tensor=tensor([[[[0.9531, 0.1133, 0.3477,  ..., 0.0273, 0.3398, 0.8359],

[...skipped tensor content...]

E                     [0.7656, 0.1016, 0.9258, 0.9336, 0.1250]]]], dtype=torch.bfloat16), device=<tt_lib.device.Device object at 0x7f1157a772f0>, in_channels=3, out_channels=8, batch_size=1, input_height=64, input_width=64, kernel_size=(5, 5), stride=[1, 1], padding=[0, 0], dilation=[1, 1], groups=1, bias_tensor=None, conv_config=None

../tt-metal/ttnn/ttnn/operations/conv2d.py:448: TypeError
```

I couldn't spot the argument mismatch, but there was still this error message.

Reply:
I make the test data smaller, so it's easier to inspect the error message.  The argument `weight_tensor` is there, and I make `kernel_size` a list instead of a tuple, but we still get the same kind of error.
```
E           TypeError: conv2d(): incompatible function arguments. The following argument types are supported:
E               1. (*, input_tensor: tt_lib.tensor.Tensor, weight_tensor: tt_lib.tensor.Tensor, device: tt_lib.device.Device, in_channels: int, out_channels: int, batch_size: int, input_height: int, input_width: int, kernel_size: List[int[2]], stride: List[int[2]], padding: List[int[2]], dilation: List[int[2]], groups: int, bias_tensor: Optional[tt_lib.tensor.Tensor] = None, conv_config: Optional[ttnn::operations::conv2d::Conv2dConfig] = None) -> Tuple[tt_lib.tensor.Tensor, int, int, tt_lib.tensor.Tensor, Optional[tt_lib.tensor.Tensor]]
E           
E           Invoked with: kwargs: input_tensor=tensor([[[[0.0781, 0.5547, 0.9531, 0.1797, 0.0039],
E                     [0.1055, 0.0312, 0.2148, 0.6719, 0.3086],
E                     [0.4609, 0.9062, 0.0000, 0.1680, 0.6328],
E                     [0.3594, 0.6914, 0.4883, 0.4961, 0.0898],
E                     [0.4766, 0.7266, 0.4219, 0.8672, 0.5352]],
E           
E                    [[0.6133, 0.0898, 0.7812, 0.4453, 0.0273],
E                     [0.2148, 0.4375, 0.6523, 0.4375, 0.9805],
E                     [0.9766, 0.2773, 0.0430, 0.5625, 0.8867],
E                     [0.7227, 0.1562, 0.1289, 0.3438, 0.7617],
E                     [0.3555, 0.9648, 0.9453, 0.5547, 0.5117]]]], dtype=torch.bfloat16), weight_tensor=tensor([[[[0.8555, 0.1055, 0.8086],
E                     [0.4805, 0.9766, 0.4883],
E                     [0.6133, 0.4727, 0.6641]],
E           
E                    [[0.3906, 0.5586, 0.8828],
E                     [0.5547, 0.7227, 0.9844],
E                     [0.1562, 0.9609, 0.2969]]],
E           
E           
E                   [[[0.9219, 0.9531, 0.5781],
E                     [0.2539, 0.9766, 0.1172],
E                     [0.3008, 0.2266, 0.6680]],
E           
E                    [[0.4336, 0.4297, 0.6641],
E                     [0.1836, 0.5664, 0.1680],
E                     [0.8828, 0.1875, 0.9531]]],
E           
E           
E                   [[[0.5820, 0.3477, 0.5078],
E                     [0.6797, 0.4297, 0.3945],
E                     [0.4844, 0.4258, 0.5820]],
E           
E                    [[0.6680, 0.2812, 0.0898],
E                     [0.3672, 0.7148, 0.5430],
E                     [0.0625, 0.9883, 0.6797]]],
E           
E           
E                   [[[0.7930, 0.9023, 0.5312],
E                     [0.9805, 0.6562, 0.0078],
E                     [0.2773, 0.2070, 0.4258]],
E           
E                    [[0.9258, 0.4766, 0.5312],
E                     [0.2852, 0.7344, 0.8047],
E                     [0.6094, 0.5352, 0.8477]]]], dtype=torch.bfloat16), device=<tt_lib.device.Device object at 0x7fda4f2a91f0>, in_channels=2, out_channels=4, batch_size=1, input_height=5, input_width=5, kernel_size=[3, 3], stride=[1, 1], padding=[0, 0], dilation=[1, 1], groups=1, bias_tensor=None, conv_config=None

../tt-metal/ttnn/ttnn/operations/conv2d.py:448: TypeError
```

Reply:
You are trying to call this function with Torch Tensors, not TT-NN Tensors.

See how dtype is `torch.bfloat16`

Reply:
See
https://github.com/tenstorrent/pytorch2.0_ttnn/blob/main/torch_ttnn/passes/lowering/add_data_move_pass.py

Reply:
Got it!  We got to register data movement for `ttnn.conv2d` first.  Now we move on to a new error message.
```
E           RuntimeError: TT_THROW @ ../ttnn/cpp/ttnn/operations/core.hpp:67: tt::exception
E           info:
E           Unable to reshape a tensor in TILE_LAYOUT to non-tile height and width! Please convert the tensor to ROW_MAJOR_LAYOUT first.
E           backtrace:
E            --- /home/ubuntu/jdh8/tt-metal/ttnn/ttnn/_ttnn.so(+0xe3ff1) [0x7f547cadbff1]
E            --- ttnn::operations::conv2d::shard_or_reshard_tensor_if_required(tt::tt_metal::Device&, tt::tt_metal::Tensor const&, ttnn::operations::conv2d::Conv2dConfig const&, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int)
E            --- ttnn::operations::conv2d::conv2d(tt::tt_metal::Tensor const&, tt::tt_metal::Tensor const&, tt::tt_metal::Device&, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, std::__1::array<unsigned int, 2ul>, std::__1::array<unsigned int, 2ul>, std::__1::array<unsigned int, 2ul>, std::__1::array<unsigned int, 2ul>, unsigned int, std::__1::optional<tt::tt_metal::Tensor const>, std::__1::optional<ttnn::operations::conv2d::Conv2dConfig const>)
E            --- /home/ubuntu/jdh8/tt-metal/ttnn/ttnn/_ttnn.so(+0x266641) [0x7f547cc5e641]
E            --- /home/ubuntu/jdh8/tt-metal/ttnn/ttnn/_ttnn.so(+0x2fbd4e) [0x7f547ccf3d4e]
E            --- /home/ubuntu/jdh8/tt-metal/python_env/bin/python3(PyCFunction_Call+0x59) [0x5f58f9]
E            --- /home/ubuntu/jdh8/tt-metal/python_env/bin/python3(_PyObject_MakeTpCall+0x29e) [0x5f64ce]
E            --- /home/ubuntu/jdh8/tt-metal/python_env/bin/python3(_PyEval_EvalFrameDefault+0x62fd) [0x5716ed]
E            --- /home/ubuntu/jdh8/tt-metal/python_env/bin/python3(_PyEval_EvalCodeWithName+0x26a) [0x569dfa]
E            --- /home/ubuntu/jdh8/tt-metal/python_env/bin/python3(_PyFunction_Vectorcall+0x393) [0x5f5e83]
```

Currently I use this set of tensor shapes for testing.
```py
@pytest.mark.parametrize(
    "input_shapes",
    [((1, 2, 5, 5), (4, 2, 3, 3))],
)
```

- - -

I wondered if the shapes have something to do with this error.  I tried setting all channels to 1 and got a segfault.
```py
@pytest.mark.parametrize(
    "input_shapes",
    [((1, 1, 5, 5), (1, 1, 3, 3))],
)
```

However, the segfault did not directly originate from C++ code in `tt-metal`.  I found this rather fishy.  :fish:

```
tests/lowering/misc/test_conv_2d.py Fatal Python error: Segmentation fault

Thread 0x00007f1d13fff700 (most recent call first):
  File "/usr/lib/python3.8/threading.py", line 306 in wait
  File "/usr/lib/python3.8/threading.py", line 558 in wait
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/tqdm/_monitor.py", line 60 in run
  File "/usr/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/lib/python3.8/threading.py", line 890 in _bootstrap

Current thread 0x00007f1e1aae7740 (most recent call first):
  File "/home/ubuntu/jdh8/tt-metal/ttnn/ttnn/operations/conv2d.py", line 448 in conv2d
  File "/home/ubuntu/jdh8/tt-metal/ttnn/ttnn/decorators.py", line 328 in __call__
  File "<eval_with_key>.19", line 7 in forward
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1520 in _call_impl
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1511 in _wrapped_call_impl
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/fx/graph_module.py", line 304 in __call__
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/fx/graph_module.py", line 738 in call_wrapped
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/_functorch/_aot_autograd/utils.py", line 81 in g
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py", line 118 in rng_functionalization_wrapper
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/_functorch/_aot_autograd/utils.py", line 105 in call_func_at_runtime_with_args
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 94 in runtime_wrapper
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/_functorch/_aot_autograd/utils.py", line 81 in g
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 901 in forward
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/_dynamo/external_utils.py", line 17 in inner
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 489 in _fn
  File "/home/ubuntu/jdh8/pytorch2.0_ttnn/tests/lowering/misc/test_conv_2d.py", line 13 in forward
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1520 in _call_impl
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1511 in _wrapped_call_impl
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 489 in _fn
  File "/home/ubuntu/jdh8/pytorch2.0_ttnn/tests/lowering/misc/test_conv_2d.py", line 30 in test_conv_2d
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/python.py", line 195 in pytest_pyfunc_call
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/python.py", line 1789 in runtest
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/runner.py", line 167 in pytest_runtest_call
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/runner.py", line 260 in <lambda>
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/runner.py", line 339 in from_call
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/runner.py", line 259 in call_runtest_hook
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/runner.py", line 220 in call_and_report
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/runner.py", line 131 in runtestprotocol
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/runner.py", line 112 in pytest_runtest_protocol
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/main.py", line 349 in pytest_runtestloop
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/main.py", line 324 in _main
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/main.py", line 270 in wrap_session
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/main.py", line 317 in pytest_cmdline_main
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/config/__init__.py", line 167 in main
  File "/home/ubuntu/jdh8/tt-metal/python_env/lib/python3.8/site-packages/_pytest/config/__init__.py", line 190 in console_main
  File "/home/ubuntu/jdh8/tt-metal/python_env/bin/pytest", line 8 in <module>
```